File: src/fragment/EntityFragment.ts
import { Text } from "pixi.js";
import Fragment, { FragmentType, register } from "./Fragment";
import { StructEndecBuilder, BuiltInEndecs, uuidToString, uuidFromString, PrimitiveEndecs } from 'KEndec';


export default class EntityFragment extends Fragment {
    uuid: string;
    name: string;

    constructor(uuid: string, name: string) {
        super();

        this.uuid = uuid;
        this.name = name;
    }

    override toString(): string {
        const match = this.name.match(/"([^"]+)"/);
        const name = match ? match[1] : "entity";

        return name;
    }

    override type(): FragmentType<EntityFragment> {
        return ENTITY;
    }
}

const ENTITY = register("entity", 0x338888, 
    StructEndecBuilder.of2(
        BuiltInEndecs.Uuid
            .xmap(uuidToString, uuidFromString)
            .fieldOf("uuid", (fragment: EntityFragment) => fragment.uuid),
        PrimitiveEndecs.STRING.fieldOf("name", (fragment: EntityFragment) => fragment.name),
        (uuid, name) => new EntityFragment(uuid, name)
    )
);

File: src/jsMain/kotlin/me/maplesyrum/tricksterstudio/spell/fragment/EntityFragment.kt
package me.maplesyrum.tricksterstudio.spell.fragment

import com.benasher44.uuid.Uuid
import com.benasher44.uuid.uuidFrom
import tree.maple.kendec.StructEndec
import tree.maple.kendec.impl.StructEndecBuilder
import tree.maple.kendec.PrimitiveEndecs
import tree.maple.kendec.impl.BuiltInEndecs


class EntityFragment(val uuid: String, val name: String) : Fragment() {
    override fun type(): FragmentType<*> {
        return FragmentType.ENTITY
    }

    override fun toString(): String {
        val match = Regex("\"([^\"]+)\"").find(name)
        val entityName = match?.groupValues?.get(1) ?: "entity"
        return entityName
    }

    companion object {
        val ENDEC: StructEndec<EntityFragment> = StructEndecBuilder.of(
            BuiltInEndecs.Uuid
                .xmap({ uuid: Uuid -> uuid.toString() }, { str: String -> uuidFrom(str) })
                .fieldOf("uuid", EntityFragment::uuid),
            PrimitiveEndecs.STRING.fieldOf("name", EntityFragment::name),
            { uuid: String, name: String -> EntityFragment(uuid, name) }
        )
    }
}
